src/db/traits.rs

```diff
@@ -0,0 +1,144 @@
+use revm::{
+    db::{states::bundle_state::BundleRetention, BundleState, State},
+    primitives::B256,
+    Database,
+};
+use std::{collections::BTreeMap, convert::Infallible, sync::Arc};
+
+/// Abstraction trait covering types that accumulate state changes into a
+/// [`BundleState`].
+///
+#[cfg_attr(
+    feature = "concurrent-db",
+    doc = "This trait is implemented for [`ConcurrentState`] in the `concurrent-db` feature."
+)]
+#[cfg_attr(
+    feature = "concurrent-db",
+    doc = "[`ConcurrentState`]: crate::db::sync::ConcurrentState"
+)]
+pub trait StateLike {
+    /// Set the state clear flag. See [`State::set_state_clear_flag`].
+    fn set_state_clear_flag(&mut self, flag: bool);
+
+    /// Merge transitions into the bundle. See [`State::merge_transitions`].
+    fn merge_transitions(&mut self, retention: BundleRetention);
+
+    /// Take the bundle. See [`State::take_bundle`].
+    fn take_bundle(&mut self) -> BundleState;
+
+    /// Set the block hashes, overriding any existing values, and inserting any
+    /// absent values.
+    fn set_block_hashes(&mut self, block_hashes: &BTreeMap<u64, B256>);
+}
+
+impl<Db: Database> StateLike for State<Db> {
+    fn set_state_clear_flag(&mut self, flag: bool) {
+        State::set_state_clear_flag(self, flag)
+    }
+
+    fn merge_transitions(&mut self, retention: BundleRetention) {
+        State::merge_transitions(self, retention)
+    }
+
+    fn take_bundle(&mut self) -> BundleState {
+        State::take_bundle(self)
+    }
+
+    fn set_block_hashes(&mut self, block_hashes: &BTreeMap<u64, B256>) {
+        self.block_hashes.extend(block_hashes)
+    }
+}
+
+/// Abstraction trait covering types that can attempt to accumulate state, but
+/// where accumulation may fail.
+///
+/// This is a fallible version of [`StateLike`].
+pub trait SharedStateLike: Sync {
```

Comment:
how does the `Shared` prefix correlate to the description? I think this is also implying it's meant to be shared, but it's unclear how?